# Statistics and Data Science: Python Basics

Source: Adapted from Boris Thurm Notebook, Datascience and statistics (EPFL)

<img src='https://www.agent-x.com.au/wp-content/uploads/2011/06/Perfect-Programmer-dfe194b-e8d3b11-b960bd5.jpg' width="400">

Source: [Agent-X Comics - Perfect Programming](https://www.agent-x.com.au/comic/perfect-programming/)

## Contenus

- [Variables](#Variables)
- [Data Type](#Data_Type)
  - [Text data](#Text_data)
  - [Numeric data](#Numeric_data)
  - [Casting and type conversion](#Casting-and-type-conversion)
- [Operators](#Operators)
  - [Arithmetic Operators](#arithmetic-operators)
    - [Operations on integers](#operating-int)
    - [Operations on floats](#operating-float)
    - [Operations on strings](#operating-str)
    - [Order of operations](#operating-order)
  - [Assignment Operators](#assignment-operators)
  - [Comparison Operators and Boolean](#comparison-operators)
  - [Identity Operators](#identity)
  - [Logical Operators](#logical)
- [Conditionals](#Conditionals)
- [Collection of elements](#collection)
  - [Lists](#Lists)
  - [Tuples](#Tuples)
  - [Conversion](#Conversion)
  - [Indexing](#Indexing)
  - [Slicing](#Slicing)
  - [Membership operators](#membership-operators)
  - [Mutability](#Mutability)
  - [Methods for lists and tuples](#methods)


## Variables <a class="anchor" id="Variables"></a>

Que vous programmiez en Python ou dans presque n'importe quel autre langage, vous travaillerez avec des **variables**. Les variables sont des conteneurs permettant de stocker des valeurs de données. Nous parlerons plus tard des **objets**, mais une variable, comme tout en Python, est un objet. Les éléments suivants peuvent être des propriétés d'une variable :  

1. Le **type** de variable. Par exemple, est-ce un entier, comme `2`, ou une chaîne de caractères, comme `'Hello, world.'` ?  
2. La **valeur** de la variable.  

En fonction du **type** de la variable, vous pouvez effectuer différentes opérations sur elle et sur d'autres variables de type similaire.  

Une variable est créée au moment où vous lui assignez une **valeur** pour la première fois :

In [ ]:
a = 3

Remarquez que lorsque vous assignez une valeur à une variable, il n'y a pas de sortie visible. Cependant, si nous demandons maintenant `a`, sa valeur sera affichée :

In [ ]:
a

3

Faites attention, lorsque vous assignez une variable déjà affectée, sa valeur sera écrasée !

In [ ]:
a = 7
a

7

<span style='color:blue'> **Conseils :** </span>

- Utilisez des noms explicites pour vos variables : vous et toute personne lisant votre code devez pouvoir comprendre ce que représente la variable.
- Le conseil le plus précieux en programmation : utilisez une convention de nommage (définissez des règles pour nommer tout et respectez-les) ! Voici mes règles :
    - Utilisez uniquement des lettres minuscules
    - Reliez les mots avec `_` uniquement (pas de `-`)
    - Commencez par des termes généraux, puis ajoutez des termes différenciants (ex. `ventes_total`, `ventes_diff`, `ventes_log`, etc.)
- Vous pouvez - en fait, vous devriez - commenter votre code en utilisant `#` pour expliquer ce que vous faites.


Python fait la différence entre **minuscules et majuscules**:

In [ ]:
# Defining "A" will not overwrite "a"
A = "What a wonderful day!"
print(A)
print(a)

What a wonderful day!
3


## Type de données <a class="anchor" id="Data_Type"></a>

Les variables peuvent stocker des données de types différents, et différents types peuvent accomplir différentes tâches. La fonction intégrée de Python `type()` permet de déterminer le type de certaines données ou variables.

### Données en texte <a class="anchor" id="Text_data"></a>

Le premier type de données que nous avons rencontré est le **Texte**, comme `"Hello, world"`. En langage Python, les données de type Texte sont appelées **string**. Lorsqu'on demande le type de `"Hello, world"`, la fonction `type()` renverra `str`, la version abrégée de string :

In [ ]:
type("Hello world")

str

Notez qu'il existe plusieurs façons de définir des chaînes de caractères. Vous pouvez utiliser des guillemets simples ou doubles. Par exemple, `'Ceci est une chaîne'` et `"Ceci est une chaîne"` sont équivalents. Vous pouvez également utiliser des triples guillemets pour étendre une chaîne sur plusieurs lignes :

In [ ]:
my_str = '''Triple quotes allows...
to extend strings over multiple lines...
An you can continue'''

print(my_str)

Triple quotes allows...
to extend strings over multiple lines...
An you can continue


### Données numériques <a class="anchor" id="Numeric_data"></a>

Il existe trois types numériques en Python :
- entier `int`
- réel `float`
- complexe `complex`


**Integer** (**int**)  est un nombre entier, positif ou négatif, sans décimales et de longueur illimitée :

In [ ]:
type(4)

str

**Float** signifie "floating point number" (ie, "nombre à virgule flottante") et désigne un nombre, positif ou négatif, contenant une ou plusieurs décimales.

In [ ]:
type(3.9)

float

Notez que vous pouvez également utiliser la notation scientifique avec `e` :


In [ ]:
type(9.5e-8)

float

Enfin, vous pouvez définir un nombre **complexe**. Notez qu'en Python, la partie imaginaire est définie par `j` :


In [ ]:
type(1+2j)

complex

Faites attention lorsque vous définissez et manipulez des données : `3.9` est un float, mais `'3.9'` est une chaîne de caractères !

### "Casting" and type conversion

Il peut arriver que vous souhaitiez spécifier un type pour une variable. Cela peut être fait par "casting", en utilisant les fonctions de construction suivantes :

- `int()` - construit un nombre entier à partir d'un entier littéral, d'un nombre flottant (en supprimant tous les décimales) ou d'une chaîne de caractères (à condition que la chaîne représente un nombre entier)
- `float()` - construit un nombre flottant à partir d'un entier littéral, d'un nombre flottant ou d'une chaîne de caractères (à condition que la chaîne représente un nombre flottant ou un entier)
- `complex()` - construit un nombre complexe à partir d'un large éventail de types de données, y compris les chaînes de caractères, les entiers littéraux et les nombres flottants
- `str()` - construit une chaîne de caractères à partir d'un large éventail de types de données, y compris les chaînes de caractères, les entiers littéraux et les nombres flottants


In [ ]:
cast_int = int(1.7)
cast_float = float(4)
cast_complex = complex(3.5)
cast_str = str(9.2)

print(type(cast_int), cast_int)
print(type(cast_float), cast_float)
print(type(cast_complex), cast_complex)
print(type(cast_str), cast_str)

<class 'int'> 1
<class 'float'> 4.0
<class 'complex'> (3.5+0j)
<class 'str'> 9.2


Notez que lors de la conversion d'un `float` en `int`, l'interpréteur n'arrondit pas le résultat, mais prend la valeur entière inférieure (floor).


In [ ]:
print(int(2.9))
print(int(3.1))

2
3


Les fonctions `int()`, `float()`, `complex()` et `str()` sont très utiles pour convertir un type de variable en un autre. Par exemple, il arrive souvent que nous importions des données depuis un fichier texte, c'est-à-dire de nombreuses chaînes de caractères, mais nous voudrons effectuer des opérations sur des nombres :


In [ ]:
imp_str = '5.3'
conv_str = float('5.3')
print(type(imp_str), imp_str)
print(imp_str*2, "we can see this is wrong")
print(type(conv_str), conv_str)
print(conv_str*2)

<class 'str'> 5.3
5.35.3 we can see this is wrong
<class 'float'> 5.3
10.6


## Opérateurs

### Opérateurs arithmétiques <a class="anchor" id="arithmetic-operators"></a>

Les **opérateurs** vous permettent d'effectuer des opérations sur les variables, comme les additionner. Ils sont représentés par des symboles spéciaux, comme `+` et `*`. Pour l'instant, nous allons nous concentrer sur les opérateurs **arithmétiques**. Les opérateurs arithmétiques en Python sont :

| action         | opérateur |
|:--------------|:----------:|
| addition      | `+`       |
| soustraction  | `-`       |
| multiplication| `*`       |
| division      | `/`       |
| puissance     | `**`      |
| modulo        | `%`       |
| division entière | `//`  |

**Attention** : N'utilisez pas l'opérateur `^` pour élever un nombre à une puissance. Il s'agit en réalité de l'opérateur de XOR bit à bit, que nous ne couvrirons pas pour l'instant.


#### Opérations sur les entiers : <a class="anchor" id="operating-int"></a>

Essayons les opérateurs arithmétiques sur des entiers :


In [ ]:
3+5

8

In [ ]:
3-5

-2

In [ ]:
3*5

15

In [ ]:
3/5

0.6

In [ ]:
3**5

243

% correspond au modulo

In [ ]:
3%5

3

// corespond à la division "d'étage" (floor division).

In [ ]:
3//5

0

Notez que `3/5` produit un `float`, même si `3` et `5` sont des `int` :


In [ ]:
print(type(3+5))
print(type(3/5))
print(3/5)
print(int(3/5))

<class 'int'>
<class 'float'>
0.6
0


Notez que l'on ne peut pas diviser par zéro



In [ ]:
7/0

ZeroDivisionError: division by zero

#### Opérations sur les nombres à virgules (float) <a class="anchor" id="operating-float"></a>

Essayons maintenant les opérateurs arithmétiques sur des `float` :


In [ ]:
2.1 + 3.2

5.300000000000001

Attendez une minute ! Nous savons que `2.1 + 3.2 = 5.3`, mais Python affiche `5.300000000000001`. Cela est dû au fait que les nombres à virgule flottante sont stockés avec un nombre fini de bits en binaire. Il y aura toujours des erreurs d'arrondi.  

Cela signifie que, du point de vue de l'ordinateur, il ne peut pas considérer que `2.1 + 3.2` et `5.3` sont strictement égaux. C'est un point important à garder en tête lorsqu'on travaille avec des floats, comme nous le verrons plus tard.


In [ ]:
# Very very close to zero because of finite precision
5.3 - (2.1 + 3.2)

-8.881784197001252e-16

In [ ]:
2.1-3-2

-2.9

In [ ]:
2.1*3.2

6.720000000000001

In [ ]:
2.1/3.2

0.65625

In [ ]:
2.1**3.2

10.74241047739471

In [ ]:
2.1%3.2

2.1

In [ ]:
2.1//3.2

0.0

Tout fonctionne comme prévu à l'exception de la précision des nombres à virgule flottante mentionnée précédemment. Comme avant, vous ne pouvez pas diviser par zéro :


In [ ]:
7.4/0.0

ZeroDivisionError: float division by zero

Notez que vous pouvez effectuer des opérations sur des entiers et des floats. En d'autres termes, vous n'avez pas besoin de convertir les entiers en floats pour effectuer des opérations mixtes :


In [ ]:
1+6.8

7.8

#### Opérations sur les chaines de carctères (string) <a class="anchor" id="operating-str"></a>

Enfin, essayons certaines de ces opérations sur des chaînes de caractères. Oui, nous allons effectuer des opérations mathématiques sur des chaînes ! Que va-t-on obtenir ? Voyons voir...


In [ ]:
'We can '+'add strings!'

'We can add strings!'

Le résultat est intuitif : additionner des chaînes les concatène ! Et que se passe-t-il si nous essayons de soustraire des chaînes ?


In [ ]:
'Can we '-'subtract strings?'

TypeError: unsupported operand type(s) for -: 'str' and 'str'

Ah, dommage, nous ne pouvons pas soustraire des chaînes de caractères. En fait, cela a du sens : soustraire des chaînes serait étrange. Au moins, nous avons obtenu un message d'erreur clair nous expliquant que `str` et `str` sont des types d'opérandes non pris en charge pour l'opération `-`.

De même, nous ne pouvons pas effectuer de multiplication, d'exponentiation, etc., entre deux chaînes. Mais que se passe-t-il si nous multiplions une chaîne par un entier ?


In [ ]:
'cat '*5

'cat cat cat cat cat '

Wow, trois fois `'cat '` ! Cela a du sens : multiplier par un entier revient simplement à effectuer plusieurs additions, donc l'interpréteur Python concatène la chaîne plusieurs fois.


#### Priorité dans les opérations <a class="anchor" id="operating-order"></a>

L'ordre des opérations suit la convention habituelle. L'exponentiation est effectuée en premier, suivie de la multiplication, de la division, de la division entière et du modulo. Ensuite viennent l'addition et la soustraction.  

Par ordre de priorité, notre tableau des opérateurs arithmétiques est :

| Priorité | Opérateurs |
|:--------:|:----------:|
| 1        | `**`       |
| 2        | `*`, `/`, `//`, `%` |
| 3        | `+`, `-`   |

Vous pouvez également regrouper des opérations avec des parenthèses. Les opérations entre parenthèses sont toujours évaluées en premier.


<span style='color:blue'> **Conseils :** </span> *n'abusez pas* des parenthèses. Un excès de parenthèses rend votre code moins lisible et peut conduire à des erreurs. Faites confiance à l'ordre des opérations 😉


In [ ]:
1**3 + 2**3 + 3**3 + 4**3 + 5**3

225

In [ ]:
(1+2+3+4+5)**2

225

Wooow! The sum of the cubes of 1, 2, ..., 5 is equal to the square of the sum from 1 to 5. Can you demonstrate that this property is true for all *n*?

### Opérateurs d'affectation (assignment operators) <a class="anchor" id="assignment-operators"></a>

Les opérateurs d'affectation sont utilisés pour attribuer des valeurs aux variables. Nous en avons déjà rencontré un : c'est exact, l'opérateur `=` qui permet d'initialiser une variable.


In [ ]:
var = 7
print(type(var), var)

<class 'int'> 7


Maintenant, disons que nous voulons mettre à jour la valeur de notre variable `var`. Comme mentionné précédemment, vous pouvez directement écraser une variable. Vous pouvez également utiliser des opérations.  

Par exemple, supposons que nous voulions ajouter `3.9` à `var`. Vous pouvez utiliser l'opérateur `+` :


In [ ]:
var = var+3.9
print(type(var), var)

<class 'float'> 26.499999999999996


Notez que nous avons changé le type de notre variable `var` d'un `int` en un `float`.


Au lieu d'utiliser l'opérateur arithmétique `+` pour mettre à jour notre variable, il existait une manière plus efficace, en utilisant l'opérateur d'affectation `+=` :


In [ ]:
var = 7
var+= 3.9
print(var)

10.9


L'opérateur `+=` indique à l'interpréteur de prendre la valeur de `var` et d'y ajouter `3.9`, en modifiant le type de `var` de manière intuitive si nécessaire.

De manière similaire, les autres opérateurs arithmétiques ont des opérateurs d'affectation équivalents :

| Opérateur | Exemple | Équivalent à |
|:---------:|:-------:|:------------:|
| `=`  | `var = 7`  | `var = 7`  |
| `+=` | `var += 7` | `var = var + 7` |
| `-=` | `var -= 7` | `var = var - 7` |
| `*=` | `var *= 7` | `var = var * 7` |
| `/=` | `var /= 7` | `var = var / 7` |
| `**=` | `var **= 7` | `var = var ** 7` |
| `%=` | `var %= 7` | `var = var % 7` |
| `//=` | `var //= 7` | `var = var // 7` |


### Opérateurs de comparaison et Booléens <a class="anchor" id="comparison-operators"></a>

Les **opérateurs de comparaison** (également appelés **opérateurs relationnels**) sont utilisés pour comparer deux valeurs.


Commençons par évaluer si deux valeurs sont égales. Nous utilisons l'opérateur `==` :


In [ ]:
8 == 8

True

In [ ]:
print(8==9)
print(2.1 + 3.2)
print(5.3 == 2.1 + 3.2)
print(5.2 == 2 + 3.2)
print(5.2 == 2 + 3.3)

False
5.300000000000001
False
True
False


Wow ! Python a confirmé que 8 est égal à 8 mais n'est pas égal à 9 !

Attendez une minute, nous savons ce que signifient "True" et "False" en anglais, c'est-à-dire des mots qui indiquent la vérité. Nous pouvons deviner qu'ils ont la même signification en Python.  

Mais quel est leur type ? Jusqu'à présent, nous avons vu les types de données `str`, `int`, `float` et `complex`. Est-ce que `True` et `False` sont des chaînes de caractères ?  

Non ! `True` et `False` ont un type spécial, appelé `bool`, abréviation de **Boolean**.


In [ ]:
print(type(True))
print(type(False))

<class 'bool'>
<class 'bool'>


Les booléens sont associés à des valeurs numériques : `True` a la valeur `1`, et `False` a la valeur `0` :


In [ ]:
True == 1

True

In [ ]:
False == 0

True

Vous pouvez même effectuer des opérations arithmétiques sur des booléens. Le résultat sera un `int` :


In [ ]:
sum_bool = True + False

print(type(sum_bool), sum_bool)

<class 'int'> 1


D'accord, maintenant que nous comprenons ce que sont les booléens, testons cela avec des floats :


In [ ]:
5.3 == 5.3

True

Comme prévu. Encore une fois :


In [ ]:
2.1+3.2 == 5.3

False

Comme prévu... Attendez, quoi ?! Comment se fait-il que `2.1 + 3.2` ne soit pas `5.3` ?  

Eh bien, souvenez-vous, il y avait des erreurs d'arrondi lors de l'addition de `2.1` et `3.2`. C'est le problème de l'arithmétique en virgule flottante.  

Notez que les nombres flottants pouvant être représentés exactement en binaire ne rencontrent pas ce problème :


In [ ]:
2.2+3.2 == 5.4

True

Malheureusement, ce comportement est imprévisible, donc **n'utilisez jamais l'opérateur `==` avec des `float`**.


La comparaison ne se limite pas à l'égalité. Voici les autres opérateurs de comparaison :

| Anglais | Python |
|:-------|:------:|
| est égal à | `==` |
| est différent de | `!=` |
| est supérieur à | `>` |
| est inférieur à | `<` |
| est supérieur ou égal à | `>=` |
| est inférieur ou égal à | `<=` |

Essayons-les !


In [ ]:
-1 > 6

False

In [ ]:
4 <= 4

True

Nous pouvons même chaîner des opérateurs de comparaison :


In [ ]:
1<2<3

True

Cependant, même si c'est légal, ne mélangez pas la direction des opérateurs de comparaison :


In [ ]:
1 < 3 > 2

True

Voyez, les opérateurs de comparaison chaînés vérifient la relation élément par élément. Dans l'exemple ci-dessus, cela signifie que `1` et `2` ne sont pas comparés.


Enfin, nous pouvons utiliser des opérateurs de comparaison sur des chaînes de caractères :


In [ ]:
'Federer' > 'Nadal'

False

Attendez, quoi ?! Python est devenu fou ! Je veux dire, Python n'a jamais vu un match de tennis, alors comment peut-il comparer des joueurs de tennis ?  

Eh bien, en réalité, il ne le fait pas. Il compare les caractères des chaînes de caractères.

Comment ? En Python, les caractères sont encodés avec [Unicode](https://en.wikipedia.org/wiki/Unicode). Il s'agit d'une bibliothèque standardisée regroupant des caractères de nombreuses langues du monde entier, contenant plus de 100 000 caractères.  

Chaque caractère possède un numéro unique qui lui est attribué. Nous pouvons accéder à ce numéro en utilisant la fonction intégrée `ord()` de Python.


In [ ]:
ord('a')

97

Les opérateurs relationnels appliqués aux caractères comparent les valeurs que retourne la fonction `ord()`. Ainsi, utiliser un opérateur relationnel sur `'a'` et `'b'` revient à comparer `ord('a')` et `ord('b')`.  

Lorsqu'on compare des chaînes de caractères, l'interpréteur commence par comparer le premier caractère de chaque chaîne. S'ils sont égaux, il passe au second caractère, et ainsi de suite.  

Ainsi, si `'Federer' > 'Nadal'` renvoie `False`, c'est parce que `ord('F') < ord('N')`. Nous sommes rassurés, mais le débat n'est toujours pas tranché... 😉

Notez qu'une conséquence de ce mécanisme est que tester l'égalité de deux chaînes signifie que **tous** les caractères doivent être identiques. C'est le cas d'utilisation le plus courant des opérateurs relationnels avec les chaînes.


### Opérateurs d'identité <a class="anchor" id="identity"></a>

Les **opérateurs d'identité** sont utilisés pour comparer des objets, non pas pour savoir s'ils sont égaux, mais pour vérifier s'ils sont réellement le même objet, c'est-à-dire s'ils occupent la même place en mémoire.  

Les deux opérateurs d'identité sont :

| Anglais | Python |
|:-------|:------:|
| est le même objet | **`is`** |
| n'est pas le même objet | **`is not`** |

C'est exact, ces opérateurs sont pratiquement identiques à l'anglais !  

Voyons ces opérateurs en action pour bien comprendre la différence entre `==` et `is`.  

Utilisons l'opérateur **`is`** pour examiner la manière dont Python stocke les variables en mémoire, en commençant par les `float`.


In [ ]:
a = 6.1
b = 6.1

a == b, a is b

(True, False)

Voyez, `a` et `b` ont la même valeur, donc l'opérateur `==` renvoie `True`.  

Cependant, ils ne sont pas le même objet car ils sont stockés à des emplacements différents en mémoire, ce qui explique pourquoi l'opérateur `is` renvoie `False`.

Ils peuvent occuper le même emplacement en mémoire si nous effectuons une affectation `b = a` :


In [ ]:
a = 6.1
b = a

a == b, a is b

(True, True)

Parce que nous avons affecté `b = a`, ils ont nécessairement la même valeur (immuable).  

Les deux variables occupent également le même emplacement en mémoire pour des raisons d'efficacité.  

Ainsi, les opérateurs `==` et `is` renvoient tous les deux `True`.


Cependant, si nous réaffectons la valeur de `a`, l'interpréteur place `a` dans un nouvel espace en mémoire, donc `a` et `b` ne sont plus le même objet :


In [ ]:
a = 6.1
b = a
a = 8.5

a == b, a is b

(False, False)

La même discussion est valable pour la plupart des `int` et `str`. Pourquoi "la plupart" et non "tous" ?

Pour les entiers compris entre `-5` et `256`, Python utilise un mécanisme appelé **mise en cache des entiers** (*integer caching*), ce qui signifie que ces entiers occupent le même emplacement en mémoire.  

Ce mécanisme de mise en cache ne s'applique pas aux entiers plus petits que `-5` ou plus grands que `256` :


In [ ]:
a = 93
b = 93
c = 708
d = 708

a is b, c is d

(True, False)

De manière similaire, Python applique parfois un mécanisme appelé [**internement des chaînes**](https://en.wikipedia.org/wiki/String_interning), qui permet un traitement des chaînes (parfois très) efficace.  

Le fait que deux chaînes occupent le même emplacement en mémoire dépend de leur contenu :


In [ ]:
a = 'Hello'
b = 'Hello'
c = 'Hello world!'
d = 'Hello world!'

a is b, c is d

(True, False)

Vous n'avez généralement pas besoin de vous soucier de la mise en cache et de l'internement pour les variables **immuables**.  

Immuable signifie qu'une fois créées, ces variables ne peuvent pas être modifiées. Si nous changeons leur valeur, la variable obtient un nouvel emplacement en mémoire.  

Toutes les variables que nous avons rencontrées jusqu'à présent (`int`, `float`, `complex` et `str`) sont immuables.


### Opérateurs logiques <a class="anchor" id="logical"></a>

Les **opérateurs logiques** peuvent être utilisés pour combiner des opérateurs relationnels et d'identité. Python possède trois opérateurs logiques :

| Logique | Python |
|:-------|:------:|
| ET     | `and`  |
| OU     | `or`   |
| NON    | `not`  |

L'opérateur `and` signifie que si les deux opérandes sont `True`, alors le résultat est `True`.  

L'opérateur `or` renvoie `True` si *au moins un* des opérandes est `True`.  

Enfin, l'opérateur `not` inverse le résultat logique.


In [ ]:
True and True

True

In [ ]:
True and False

False

In [ ]:
False and False

False

In [ ]:
True or False

True

In [ ]:
not False

True

In [ ]:
not False and True

True

In [ ]:
not (True and False)

True

In [ ]:
1==1 and 2==3

False

In [ ]:
1==1 or 2==3

True

Notez qu'il est important de spécifier l'ordre de vos opérations, en particulier lorsque vous utilisez l'opérateur `not`.

Notez également que

```python
a < b < c
```

est équivalent à

```python
(a < b) and (b < c)
```

Avec ces nouveaux types d'opérateurs, nous pouvons maintenant établir une table plus complète de la **priorité des opérateurs**.

| priorité | opérateurs |
|:--------:|:----------:|
| 1        | `**`       |
| 2        | `*`, `/`, `//`, `%` |
| 3        | `+`, `-`   |
| 4        | `<`, `>`, `<=`, `>=` |
| 5        | `==`, `!=` |
| 6        | `=`, `+=`, `-=`, `*=`, `/=`, `**=`, `%=`, `//=` |
| 7        | `is`, `is not` |
| 8        | `and`, `or`, `not` |


## Conditionnels

Les **conditionnelles** sont utilisées pour indiquer à votre ordinateur d'exécuter un ensemble d'instructions en fonction du fait qu'un booléen soit `True` ou non. En d'autres termes, nous disons à l'ordinateur :

```python
if quelque chose est vrai :
    faire tâche a
sinon :
    faire tâche b
```

En réalité, la syntaxe en Python est presque exactement la même. Comme toujours, un exemple parle plus fort.

Nous allons étudier la condition de coopération dans le [problème d'action collective](https://en.wikipedia.org/wiki/Collective_action_problem), également appelé dilemme social. Dans une telle situation, tous les individus seraient mieux lotis en coopérant mais échouent à le faire en raison des intérêts conflictuels entre eux, ce qui décourage l'action commune (voir l'illustration ci-dessous). De nombreuses questions environnementales prennent la forme d'un dilemme social. Par exemple, le recyclage nécessite du temps et des efforts, mais réduit la consommation de matériaux si largement adopté. L'achat d'un véhicule électrique est coûteux, mais réduit les polluants atmosphériques, associés à diverses maladies respiratoires et cardiovasculaires, ainsi que les émissions de gaz à effet de serre, responsables du changement climatique.


Nous supposerons que les individus ont des préférences *homo moralis* : ils prennent en compte non seulement leur profit égoïste, mais aussi ce qui se passe lorsque tous les autres font la même action. Le poids de l'égoïsme et de la moralité dépend du degré de moralité de l'individu. La littérature économique récente a démontré qu'une telle préférence procure un avantage évolutif (voir par exemple, Alger & Weibull, 2013).

Nous pouvons démontrer que les individus *homo moralis* coopèrent dans un dilemme social (par exemple, réalisent une action pro-environnementale) lorsque leur bénéfice social pondéré par leur degré de moralité est supérieur à leur coût individuel de l'action pondéré par leur degré d'égoïsme.


<img src='https://i.postimg.cc/44HkDp79/Social-Dilemma.png' width="800">

D'accord, assez de mots, évaluons si un *homo moralis* donné coopère.

*Référence*
Alger, I., & Weibull, J. W. (2013). Homo moralis—preference evolution under incomplete information and assortative matching. Econometrica, 81(6), 2269-2302. [DOI: 10.3982/ECTA10637](https://doi.org/10.3982/ECTA10637)



In [ ]:
cost = 1            # individual cost
benefit = 3         # social benefit
kappa = 0.5         # degree of morality

# condition for cooperation:
# social benefit times kappa is greater than individual cost times (1-kappa)

if benefit*kappa >= cost*(1-kappa):
    print('The individual cooperates!')

The individual cooperates!


Youhouuu, bonne nouvelle pour la nature, l'individu effectue une action respectueuse de l'environnement !

Maintenant, révisons la syntaxe de la déclaration `if`. L'expression booléenne, `benefit*kappa >= cost*(1-kappa)`, est appelée la **condition**. Si elle est `True`, l'instruction indentée en dessous est exécutée. Dans ce cas, nous affichons la chaîne `'L'individu coopère !'`. De plus, n'oubliez pas le `:` à la fin de la déclaration `if` !

Cela soulève un aspect très important de la syntaxe Python : <span style="color: dodgerblue; font-weight: bold;">L'indentation compte.</span> Toutes les lignes avec le même niveau d'indentation seront évaluées ensemble.


In [ ]:
if benefit*kappa >= cost*(1-kappa):
    print('The individual cooperates!')
    print('Same level of intentation, so still printed!')

The individual cooperates!
Same level of intentation, so still printed!


Il est important de noter que l'identation (note: l'indentation représente un ou plusieurs espaces au début d'une ligne de code) est important. Dans l'exemple ci-dessous, l'identation de la deuxième et celle de la troisième ligne de code ne sont pas les mêmes, ce qui pose un problème.

In [ ]:
if benefit*kappa >= cost*(1-kappa):
    print('The individual cooperates!')
      print('Same level of intentation, so still printed!')

IndentationError: unexpected indent (<ipython-input-3-92cd61dd0ddf>, line 3)

Maintenant, que se passe-t-il si la condition est `False` ? Essayons avec un individu ayant un degré de moralité `kappa=0`, c'est-à-dire le fameux *homo oeconomicus* entièrement égoïste :


In [ ]:
kappa = 0                # degree of morality

# condition for cooperation:
# social benefit times kappa is greater than individual cost times (1-kappa)

if benefit*kappa >= cost*(1-kappa):
    print('The individual cooperates!')

Rien ne s'est passé. Cela est dû au fait que nous n'avons pas indiqué à Python ce qu'il devait faire si la condition était évaluée comme `False`. Nous pouvons ajouter cela avec une **clause** `else` dans la condition.


In [ ]:
kappa = 0      # degree of morality

# condition for cooperation:
# social benefit times kappa is greater than individual cost times (1-kappa)

if benefit*kappa >= cost*(1-kappa):
    print('The individual cooperates!')
else:
    print('What a shame, the individual does not cooperate...')

What a shame, the individual does not cooperate...


Nous pouvons évaluer plusieurs conditions en utilisant une clause `elif`. Par exemple, disons que nous avons deux individus, Edoardo et Quentin, et que nous voulons vérifier s'ils coopèrent tous les deux, si un seul d'entre eux coopère, ou s'ils ne se soucient pas tous les deux de l'environnement :


In [ ]:
kappa_1 = 0.2     # degree of morality of the first individual
kappa_2 = 0.3     # degree of morality of the second individual

# condition for cooperation:
# social benefit times kappa is greater than individual cost times (1-kappa)

if benefit*kappa_1 >= cost*(1-kappa_1) and benefit*kappa_2 >= cost*(1-kappa_2):
    print('Both individual cooperates!')
elif benefit*kappa_1 < cost*(1-kappa_1) and benefit*kappa_2 < cost*(1-kappa_2):
    print('What a shame, nobody cooperates...')
else:
    print('Only one individual cooperates')

Only one individual cooperates


Il semble qu'aucun des deux, Edoardo ou Quentin, ne coopère... Quoi qu'il en soit, notez l'utilisation de l'opérateur logique `and` en plus des opérateurs de comparaison `>=` et `<`.


## Collection d'éléments <a class="anchor" id="collection"></a>

Nous allons maintenant explorer deux types de données importants en Python : les listes et les tuples. Ce sont tous deux des séquences d'objets. Tout comme une chaîne de caractères est une séquence (c'est-à-dire, une collection ordonnée) de caractères, les listes et les tuples sont des séquences d'objets arbitraires, appelés éléments. Ce sont des moyens de créer un objet unique qui contient de nombreux autres objets.


### Listes

Les **listes** sont utilisées pour stocker plusieurs éléments dans une seule variable. Nous créons des listes en mettant des valeurs ou des expressions Python à l'intérieur de **crochets**, séparées par des **virgules** :


In [ ]:
my_list = [2, 3.7, 4+5j, 'dog']
print(type(my_list),my_list)

<class 'list'> [2, 3.7, (4+5j), 'dog']


Notez que le type d'une liste est... une `list` ! De plus, toute expression Python peut faire partie d'une liste, y compris une autre liste :


In [ ]:
my_list2 = [2, 3.7, 4+5j, 'dog', [0,'Hi!']]
print(my_list2)

[2, 3.7, (4+5j), 'dog', [0, 'Hi!']]


Vous pouvez également effectuer des opérations à l'intérieur d'une liste. Dans ce cas, les opérations sont évaluées :


In [ ]:
my_list3 = [8+9, 8-9, 8*9]
print(my_list3)

['8+9', -1, 72]


Que se passe-t-il lorsque vous effectuez des opérations sur des listes ? Découvrons-le !


Les opérateurs sur les listes se comportent de manière similaire à ceux des chaînes de caractères. L'opérateur `+` sur les listes signifie la concaténation de listes.


In [ ]:
[1,2,3]+[4,5,6]

[1, 2, 3, 4, 5, 6]

L'opérateur `*` sur les listes signifie la réplication et la concaténation des listes.


In [ ]:
[1,2,3]*3

[1, 2, 3, 1, 2, 3, 1, 2, 3]

### Tuples

Tout comme les listes, les **tuples** sont utilisés pour stocker plusieurs éléments dans une seule variable et sont IMMUTABLES (ils ne peuvent pas être modifiés). Nous créons des tuples en mettant des valeurs ou des expressions Python à l'intérieur de **parenthèses**, séparées par des **virgules** :


In [ ]:
my_tuple = (2, 3.7, 4+5j, 'dog', (0,'Hi!'))
print(type(my_tuple),my_tuple)

<class 'tuple'> (2, 3.7, (4+5j), 'dog', (0, 'Hi!'))


Le type d'un tuple est, comme vous l'avez deviné, un `tuple`. Tout comme pour les listes, toute expression Python peut faire partie d'un tuple, y compris un autre tuple, et vous pouvez également effectuer des opérations à l'intérieur des tuples :


In [ ]:
(8+9, 8-9, 8*9)

(17, -1, 72)

Soyez simplement prudent lorsque vous créez un tuple avec un seul élément : vous devez inclure une virgule après l'élément :


In [ ]:
my_tuple = (0,)
not_a_tuple = (0) # this is just the number 0 (normal use of parantheses)

type(my_tuple), type(not_a_tuple)

(tuple, int)

Les opérateurs sur les tuples fonctionnent de la même manière que pour les listes, c'est-à-dire que vous pouvez concaténer des tuples avec les opérateurs `+` et `*` :


In [ ]:
(1,2,3)+(4,5,6)

(1, 2, 3, 4, 5, 6)

In [ ]:
(1,2,3)*5

(1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3)

### Conversion

Vous pouvez convertir un `tuple` en `list` en utilisant la fonction `list()` :


De la même manière, vous pouvez convertir une `list` en `tuple` en utilisant la fonction `tuple()` :


In [ ]:
tuple_to_convert = (0,1,2,3)
converted_list = list(tuple_to_convert)

converted_list

[0, 1, 2, 3]

In [ ]:
list_to_convert = [0,1,2,3]
converted_tuple = tuple(list_to_convert)

converted_tuple

(0, 1, 2, 3)

### Indexation

Les listes et les tuples sont **ordonnés**, ce qui signifie que les éléments ont un ordre défini. Ainsi, nous pouvons accéder à un élément donné dans une liste ou un tuple. Pour ce faire, nous utilisons les **crochets**. Nous écrivons d'abord le nom de notre liste/tuple, puis entre crochets, nous indiquons l'emplacement (**index**) de l'élément désiré :


In [ ]:
list_index = [2, 3.7, 4+5j, 'dog', [0,'Hi!']]

list_index[1]

3.7

Attendez, quoi ?! Nous avons demandé le premier élément et nous avons obtenu le deuxième élément de notre liste. Est-ce que Python ne sait pas compter ? Ne vous inquiétez pas, ce comportement se produit parce que <span style='color:red'> **l'indexation en Python commence à zéro** </span>. C'est très important. (Note historique : [Pourquoi Python utilise l'indexation commençant à zéro](http://python-history.blogspot.com/2013/10/why-python-uses-0-based-indexing.html).)


In [ ]:
print(list_index[0])
print(list_index[4])

2
[0, 'Hi!']


Bien mieux !

Dans notre deuxième exemple, nous avons accédé à la liste qui se trouvait dans notre liste, c'est-à-dire une sous-liste. Une liste qui contient une autre liste s'appelle une **liste imbriquée**. La sous-liste peut également contenir une autre liste (c'est-à-dire une sous-sous-liste), et ainsi de suite. Nous pouvons indexer une sous-liste en ajoutant un autre ensemble de crochets :


In [ ]:
nested_list = [[1,2,3],[4,5,6]]

print(nested_list[0][1])
print(nested_list[1][0])

2
4


Il en va de même pour les tuples : vous pouvez indexer des tuples imbriqués avec plusieurs ensembles de crochets :


In [ ]:
nested_tuple = ((1,2,3),(4,5,6))

print(nested_tuple[0][2])
print(nested_tuple[1][1])

3
5


D'accord, maintenant nous connaissons les bases de l'indexation. Une fonctionnalité incroyable offerte par Python est **l'indexation négative**. Cela signifie simplement que nous commençons l'indexation à partir de la dernière entrée, en commençant par `-1` :


In [ ]:
list_index2 = [2, 3.7, 4+5j, 'dog']

list_index2[-1]

'dog'

L'indexation à l'envers est parfois très pratique. Récapitulons les indices directs et inverses pour les listes et les tuples :

| Élément               | 1  | 2  | 3  | 4  | 5  | 6  | 7  | 8  | 9  | 10 |
|-----------------------|----|----|----|----|----|----|----|----|----|----|
| Indices directs       | 0  | 1  | 2  | 3  | 4  | 5  | 6  | 7  | 8  | 9  |
| Indices inverses      | -10| -9 | -8 | -7 | -6 | -5 | -4 | -3 | -2 | -1 |


In [ ]:
tuple_index = (1,2,3,4,5,6,7,8,9,10)

print(tuple_index[7])
print(tuple_index[-3])

8
8


### Découpage (slicing)

Avec l'indexation, nous avons accédé à un élément donné. Comment **découper** une liste ou un tuple, c'est-à-dire extraire plusieurs éléments ? Nous pouvons utiliser les deux-points `[:]` pour cela :


In [ ]:
slicing_list = [0,1,2,3,4,5,6,7,8,9,10]

slicing_list[0:4]

[0, 1, 2, 3]

Dans l'exemple ci-dessus, nous avons extrait une liste avec les éléments de `0` à `3`, malgré le fait que nous ayons écrit `[0:4]`. En d'autres termes, le dernier élément (`4`) n'est pas inclus.

De manière générale, lorsqu'on utilise l'indexation par deux-points `[i:j]`, nous obtenons les éléments de `i` à `j-1`. Autrement dit, la plage est **inclusive du premier index et exclusive du dernier**. Si l'index final du slice est supérieur à la longueur de la séquence, le slice se termine à l'élément final. Ainsi, soyez prudent lorsque vous découpez des listes/tuples.


In [ ]:
slicing_list[3:100]

[3, 4, 5, 6, 7, 8, 9, 10]

Comme précédemment, vous pouvez utiliser des indices négatifs :


In [ ]:
slicing_list[3:-2]

[3, 4, 5, 6, 7, 8]

Lorsque `i` est plus grand que `j` lors de l'utilisation de l'indexation par deux-points `[i:j]` (en termes d'indices), nous obtenons une liste vide :


In [ ]:
slicing_list[7:-5]

[]

Jusqu'à présent, nous avons extrait des éléments consécutifs. Que faire si vous ne voulez que les nombres pairs de notre liste `[0,1,2,3,4,5,6,7,8,9,10]` ? Eh bien, vous pouvez spécifier un **stride** en utilisant un deuxième deux-points (qui donne le pas) :


In [ ]:
slicing_list[0::2]

[0, 2, 4, 6, 8, 10]

Dans l'exemple ci-dessus, `0` est l'indice de départ et `2` définit le stride, c'est-à-dire le pas. Lorsque le départ n'est pas défini, la valeur par défaut est zéro :


In [ ]:
slicing_list[::2]   #no need to specify "0" when we want to start at the first index

[0, 2, 4, 6, 8, 10]

Supposons maintenant que nous voulions les nombres impairs, comment faisons-nous ? C'est simple, il suffit de modifier la valeur de départ de notre stride !


In [ ]:
slicing_list[1::2]

[1, 3, 5, 7, 9]

Et si nous voulons les multiples de trois ? Nous modifions le stride :


In [ ]:
slicing_list[::3]

[0, 3, 6, 9]

Que dire de la valeur entre les deux deux-points ? Jusqu'à présent, nous l'avons laissée non définie. En réalité, il s'agit de l'indice de fin :


In [ ]:
slicing_list[:6:2]

[0, 2, 4]

Faisons un récapitulatif du fonctionnement de l'indexation et du découpage. La structure générale est : `[start:end:stride]`

* S'il n'y a pas de deux-points, un seul élément est retourné.
* S'il y a des deux-points, nous effectuons un découpage de la liste, et une nouvelle liste est retournée.
* Si un seul deux-points est présent, `stride` est supposé être égal à 1.
* Si `start` n'est pas spécifié, il est supposé être zéro.
* Si `end` n'est pas spécifié, on suppose que vous voulez toute la liste.
* Si `stride` n'est pas spécifié, il est supposé être 1.

Maintenant, faisons un peu de découpage fou ! Imaginez que nous voulons inverser une liste/tuple. Pouvez-vous penser à un moyen de faire cette opération en utilisant le découpage ? Eh bien, nous pouvons utiliser un `stride` négatif !


In [ ]:
slicing_list[::-1]

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]

Notez que la signification des indices "start" et "end" est un peu ambiguë lorsque vous utilisez un stride négatif. Lorsque le stride est négatif, nous effectuons toujours un découpage de start à end, mais l'ordre est inversé.


In [ ]:
slicing_list[-1:6:-2]

[10, 8]

Prenez un moment pour pratiquer le découpage, car c'est un concept très important :)


### Opérateurs d'appartenance (membership operators) <a class="anchor" id="membership-operators"></a>

Les **opérateurs d'appartenance** sont utilisés pour tester si une séquence est présente dans un objet tel qu'une liste ou un tuple. Les deux opérateurs d'appartenance sont :

| Anglais | opérateur |
|:-------:|:---------:|
| fait partie de | `in` |
| ne fait pas partie de | `not in` |

Le résultat de l'opérateur est `True` ou `False`. Voyons quelques exemples :


In [ ]:
my_list = [2, 3.7, 4+5j, 'dog', [0,'Hi!']]

2 in my_list

True

En effet, `2` est dans notre liste, c'est le premier élément. Et qu'en est-il de `'Hi!'` ?


In [ ]:
'Hi' in my_list

False

Pourquoi `'Hi!'` n'est-elle pas dans notre liste ? Eh bien, elle fait partie de notre sous-liste `[0, 'Hi!']` mais pas de notre liste "principale" `my_list`. En réalité, `my_list` contient cinq éléments, et `'Hi!'` n'en fait pas partie.

Regardons un exemple avec un tuple pour nous assurer que nous maîtrisons les opérateurs d'appartenance :


In [ ]:
my_tuple = (2, 3.7, 4+5j, 'dog', [0,'Hi!'])

'cat' not in my_tuple

True

C'est tout pour l'instant, nous utiliserons d'autres opérateurs d'appartenance plus tard...


### Mutabilité

Jusqu'à présent, il semble que `list` et `tuple` soient très similaires. Alors pourquoi y aurait-il deux types différents s'ils se comportent exactement de la même manière ? Eh bien, comme vous pouvez le deviner, ce n'est pas le cas. La différence importante entre `list` et `tuple` réside dans leur mutabilité.

Les listes sont des objets **mutables** : vous pouvez changer leurs valeurs sans créer une nouvelle liste :


In [ ]:
mutable_list = [0, 1, 2, 3, 4, 5]
mutable_list[2] = 'two'

mutable_list

[0, 1, 'two', 3, 4, 5]

`list` est le seul type de données que nous avons rencontré jusqu'à présent qui est mutable. En d'autres termes, `int`, `float`, `complex`, `str`, et `bool` sont des types **immuables**. Immuable signifie qu'une fois que les variables sont créées, leurs valeurs ne peuvent pas être modifiées. Si nous changeons la valeur, la variable obtient un nouvel emplacement en mémoire. `tuple` est également un objet immuable. Essayons la même opération que nous avons effectuée précédemment sur notre liste et voyons ce qui se passe :


In [ ]:
immutable_tuple = (0, 1, 2, 3, 4, 5)
immutable_tuple[2] = 'two'

immutable_tuple

TypeError: 'tuple' object does not support item assignment

Nous obtenons un message d'erreur et à juste titre : puisque `tuple` est immuable, il ne prend pas en charge l'affectation d'éléments.


Nous pouvons utiliser la fonction `id()` pour mieux comprendre la propriété de mutabilité. Cette fonction nous indique où en mémoire la variable est stockée. Essayons :


In [ ]:
immutable_int = 89
print(id(immutable_int))

immutable_int = 90
print(id(immutable_int))

2333706441840
2333706441872


Voyez, lorsque nous changeons la valeur de `immutable_int`, nous n'avons en réalité pas modifié sa valeur ; nous avons créé une nouvelle variable ! Les listes se comportent différemment :


In [ ]:
mutable_list = [0, 1, 2, 3, 4, 5]
print(id(mutable_list))

mutable_list[1] = 'one'
print(id(mutable_list))

2333780786688
2333780786688


C'est toujours la même liste, même si nous avons changé la valeur du deuxième élément.


À ce stade, vous vous demandez peut-être : pourquoi cela nous importe-t-il ? Eh bien, supposons que nous ayons une liste, que nous souhaitons conserver, et que nous voulons en faire une copie avec un élément qui diffère. Que se passe-t-il alors ?


In [ ]:
mutable_list = [0, 1, 2, 3, 4, 5]
mutable_list_2 = mutable_list     # copy of my_list?
mutable_list_2[0] = 'zero'

print(mutable_list, mutable_list_2)

['zero', 1, 2, 3, 4, 5] ['zero', 1, 2, 3, 4, 5]


Catastrophe ! Nous avons perdu `mutable_list` !

Que s'est-il passé ? Eh bien, affecter une liste à une variable ne copie pas la liste dans un nouvel objet, cela crée simplement une nouvelle référence au même objet. Ainsi, lorsque nous avons modifié le premier élément de `mutable_list_2`, nous avons également modifié `mutable_list` ! Ce comportement peut entraîner des bugs désagréables qui vous causeront bien des ennuis !

Existe-t-il un moyen de résoudre ce problème ? Bien sûr, il y en a un : nous pouvons utiliser le découpage ! Si les indices de début et de fin du découpage d'une liste sont omis, le découpage crée une copie de toute la liste dans un nouvel emplacement mémoire.


In [ ]:
mutable_list = [0, 1, 2, 3, 4, 5]
mutable_list_2 = mutable_list[:]
mutable_list_2[0] = 'zero'

print(mutable_list, mutable_list_2)

[0, 1, 2, 3, 4, 5] ['zero', 1, 2, 3, 4, 5]


Quel soulagement !

Nous avons vu que les tuples et les listes sont très similaires, différant essentiellement uniquement par leur mutabilité (en réalité, les différences sont plus profondes, voir par exemple une discussion ici : [article mentionné](http://www.asmeurer.com/blog/posts/tuples/)).

Alors vous pouvez vous demander : "Quand dois-je utiliser un tuple et quand dois-je utiliser une liste ?" Voici le conseil de [Justin Bois](http://bois.caltech.edu/), dont le [cours](http://justinbois.github.io/bootcamp/2022_epfl/) a fortement influencé ce carnet :

"<span style="color: dodgerblue; font-weight: bold;">
Utilisez toujours des tuples au lieu de listes, sauf si vous avez besoin de mutabilité.
</span>
Cela vous évitera bien des ennuis. Il est très facile de modifier involontairement une liste, puis une autre liste (qui est en réalité la même, mais avec un nom de variable différent) se retrouve modifiée. Cela dit, la mutabilité est souvent très utile, vous pouvez donc l'utiliser pour créer votre liste et l'ajuster selon vos besoins. Cependant, après avoir finalisé votre liste, vous devriez la convertir en tuple afin qu'elle ne puisse pas être modifiée."


### Méthodes pour listes et tuples <a class="anchor" id="methods"></a>

Nous avons précédemment effectué des opérations sur les `listes`. En utilisant le découpage, nous avons extrait des éléments des listes, copié une liste avec `[:]`, et même inversé une liste avec `[::-1]`.

Que faire si nous souhaitons ajouter un élément à la fin d'une liste ? Ou mieux encore, insérer ou supprimer un élément à une position donnée. Dans ce cas, nous pouvons utiliser des fonctions intégrées.

Nous avons déjà mentionné que les listes sont des objets. Les objets contiennent : 1) des données ; 2) des fonctions qui peuvent opérer sur ces données. Les fonctions à l'intérieur d'un objet sont appelées méthodes. Voici les méthodes intégrées que vous pouvez utiliser sur les listes :

| Méthode  | Description |
|:--------:|:-----------:|
| `append()` | Ajoute un élément à la fin de la liste |
| `clear()`  | Supprime tous les éléments de la liste |
| `copy()`   | Retourne une copie de la liste |
| `count()`  | Retourne le nombre d'éléments avec la valeur spécifiée |
| `extend()` | Ajoute les éléments d'une liste (ou tout autre objet itérable) à la fin de la liste actuelle |
| `index()`  | Retourne l'indice du premier élément avec la valeur spécifiée |
| `insert()` | Ajoute un élément à la position spécifiée |
| `pop()`    | Supprime l'élément à la position spécifiée |
| `remove()` | Supprime l'élément avec la valeur spécifiée |
| `reverse()`| Inverse l'ordre des éléments dans la liste |
| `sort()`   | Trie la liste |

Une autre fonction utile (qui n'est pas une méthode) est la fonction `len()`. Elle retourne le nombre total d'éléments dans une liste. Essayons-la !


In [ ]:
my_list = [1,2,3,4,5,6,7]
len(my_list)

7

Nous avons effectivement sept éléments dans notre liste. Maintenant, comptons combien de fois la valeur `3` apparaît dans notre liste :


In [ ]:
my_list.count(3)

1

Comme prévu, nous comptons une occurrence de la valeur `3`. Mais faisons une pause un instant. Avez-vous remarqué la syntaxe ? Nous spécifions d'abord notre liste, puis un `.`, et enfin la fonction `count()`. C'est la structure d'une méthode. Maintenant, extrayons l'indice du premier élément dont la valeur est `3`. Rappelez-vous que l'indexation commence à `0`.


In [ ]:
my_list.index(3)

2

Continuons en ajoutant des éléments et des listes d'éléments à notre liste.

In [ ]:
my_list.append(8)

my_list

[1, 2, 3, 4, 5, 6, 7, 8]

In [ ]:
my_list_2 = [9,10]
my_list.extend(my_list_2)

my_list

[1, 2, 3, 5, 6, 7, 8, 9, 10, 9, 10]

Maintenant, faisons un peu de magie : nous allons faire disparaître l'élément `4` !

In [ ]:
my_list.remove(4)
my_list

ValueError: list.remove(x): x not in list

Woow ! Faisons-le réapparaître. Encore une fois, soyez prudent, l'indexation commence à `0` :


In [ ]:
my_list.insert(3,4) #insert the value 4 at the index 3 in our list
my_list

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 9, 10]

Ta-da! Instead of making a value disappear, we can also remove an element at a given position:

In [ ]:
my_list.pop(8)
my_list

[1, 2, 3, 4, 5, 6, 7, 8, 10]

Insérons à nouveau le neuvième élément :


In [ ]:
my_list.insert(8,9) #insert the value 8 at the index 9 in our list
my_list

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

D'accord, tout est bon ! Maintenant, nous pouvons trier notre liste par ordre décroissant :


In [ ]:
my_list.reverse()
my_list

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

Et trions-la à nouveau :


In [ ]:
my_list.sort() #Alternatively, we could reverse it again!
my_list

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

Enfin, supprimons tous les éléments de notre liste :


In [ ]:
my_list.clear()
my_list

[]

`clear` crée une liste vide. En d'autres termes, notre liste n'a pas été complètement effacée de l'existence, seules ses valeurs ont été supprimées.


Nous avons effectué de belles opérations sur les listes. Et les tuples ? Les fonctions `count()`, `index()`, et `len()` fonctionnent de la même manière avec les `tuple`. Qu'en est-il des autres ? Eh bien, malheureusement, elles ne fonctionnent pas de la même façon. En effet, nous avons vu précédemment que nous ne pouvons pas modifier un tuple ou un élément donné d'un tuple, car - rappelons-le, les tuples sont immuables.

Alors, que faire lorsque vous souhaitez réellement mettre à jour votre tuple ? Devez-vous en créer un nouveau ? Non, il y a une solution ! Rappelez-vous que vous pouvez convertir un tuple en une liste à l'aide de la fonction `list()` et ensuite reconvertir votre liste en tuple à l'aide de la fonction `tuple()` :


In [ ]:
my_tuple = (0,1,2,3)
my_list = list(my_tuple)
my_list[1]='one'
my_tuple = tuple(my_list)

my_tuple

(0, 'one', 2, 3)

En effectuant la conversion en `list`, vous pouvez utiliser toutes les méthodes qui fonctionnent sur les listes, puis reconvertir en `tuple` !


De plus, nous pouvons faire d'autres choses intéressantes avec les tuples. L'une d'entre elles s'appelle **l'extraction** (unpacking), et consiste en une affectation multiple. Voyons un exemple :


In [ ]:
unpacking_tuple = (1, 2, 3)
a, b, c = unpacking_tuple

print(a, b, c)

1 2 3


Cela est utile lorsque nous voulons renvoyer plusieurs valeurs depuis une fonction et utiliser ces valeurs comme stockées dans des variables distinctes.
